File: cos/api/src/main/java/cos/api/PlayerSession.java
package cos.api;

import cos.logging.Logger;
import cos.ops.Direction;
import cos.ops.OutOp;
import cos.ops.UserOp;
import cos.ops.in.FireballEmmit;
import cos.ops.in.Login;
import cos.ops.in.Logout;
import cos.ops.in.MeleeAttack;
import cos.ops.in.Move;
import cos.ops.in.ShotEmmit;
import cos.ops.in.StopMove;
import cos.ops.out.UserPackage;
import io.vertx.core.http.ServerWebSocket;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;

class PlayerSession {
    private final AtomicInteger cid = new AtomicInteger(0);
    private final Logger log = Logger.get(getClass());
    private final Consumer<UserOp> olympus;
    private final ServerWebSocket ws;
    private final int userId;
    private volatile boolean isClosed = false;


    boolean isClosed() {
        return isClosed;
    }

    PlayerSession(ServerWebSocket ws,
                  int userId,
                  Consumer<UserOp> olympus) {

        this.ws = ws;
        this.userId = userId;
        this.olympus = olympus;
        log.info("Connected player: #" + userId);
        setupClient();

        ws.closeHandler(it -> {
            isClosed = true;
            send(new Logout(cid(), userId));
            log.info("Client socket is closing ... ");
        });

        ws.textMessageHandler(this::onRequest);
    }

    private void send(UserOp op) {
        olympus.accept(op);
    }

    private void onRequest(String msg) {
        var js = new JsonObject(msg);
        var request = parseRequest(js);
        if (request != null) send(request);
    }

    private @Nullable UserOp parseRequest(JsonObject js) {
        return switch (js.getString("op")) {
            case "move" -> {
                var dir = asDir(js.getString("dir"));
                var sight = asDir(js.getString("sight"));
                if (sight == null) sight = dir;
                if (dir == null) throw new IllegalArgumentException("Wrong move request");
                yield new Move(cid(), userId, js.getInteger("x"), js.getInteger("y"), dir, sight);
            }
            case "emmit_fireball" -> new FireballEmmit(cid(), userId);
            case "emmit_shot" -> new ShotEmmit(cid(), userId);
            case "melee_attack" -> new MeleeAttack(cid(), userId);
            case "stop_move" ->
                    new StopMove(cid(), userId, js.getInteger("x"), js.getInteger("y"), asDir(js.getString("sight")));
            default -> null;
        };

    }

    private int cid() {
        return cid.incrementAndGet();
    }

    @Nullable Direction asDir(@Nullable String raw) {
        if (raw == null) return null;
        return Direction.valueOf(raw);
    }

    void onOp(UserPackage pkg) {
        var messages = new JsonArray();

        for (Record op : pkg.ops()) {
            messages.add(JsonMapper.toJson((OutOp) op));
        }

        var clientRes = new JsonObject()
                .put("tick", pkg.tick()) //todo hardcode
                .put("time", System.currentTimeMillis() / 1000)
                .put("messages", messages);
        ws.writeTextMessage(clientRes.toString());
    }

    private void setupClient() {
        send(new Login(cid(), userId));
    }
}

File: cos/api/src/main/kotlin/cos/api/PlayerSession.kt
package cos.api

import cos.logging.Logger
import cos.ops.AnyOp
import cos.ops.Direction
import cos.ops.OutOp
import cos.ops.`in`.*
import cos.ops.out.UserPackage
import io.vertx.core.http.ServerWebSocket
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import java.util.concurrent.atomic.AtomicInteger

class PlayerSession(
    private val ws: ServerWebSocket,
    private val userId: Int,
    private val olympus: (r: AnyOp) -> Int
) {
    private val cid = AtomicInteger(0)
    private val log = Logger.get(javaClass)

    @Volatile
    var isClosed = false

    init {
        log.info("Connected player: #$userId")
        setupClient()

        ws.closeHandler {
            isClosed = true
            send(Logout(cid.incrementAndGet(), userId))
            log.info("Client socket is closing ... ")
        }

        ws.textMessageHandler(::onRequest)
    }

    private fun send(op: AnyOp) {
        olympus(op)
    }

    private fun onRequest(msg: String) {
        val js = JsonObject(msg)
        val op = parseRequest(js) ?: return
        send(op)
    }

    private fun parseRequest(js: JsonObject): AnyOp? {
        return when (js.getString("op")) {
            "move" -> {
                val dirId: String? = js.getString("dir")
                var sightId = js.getString("sight")
                if (sightId == null) sightId = dirId;
                if (dirId == null && sightId == null) throw IllegalArgumentException("Wrong move request")

                Move(
                    cid.incrementAndGet(),
                    userId,
                    js.getInteger("x"),
                    js.getInteger("y"),
                    if (dirId === null) null else Direction.valueOf(dirId),
                    Direction.valueOf(sightId)
                )
            }
            "emmit_fireball" -> {
                FireballEmmit(
                    cid.incrementAndGet(),
                    userId
                )
            }
            "emmit_shot" -> {
                ShotEmmit(
                    cid.incrementAndGet(),
                    userId
                )
            }
            "melee_attack" -> MeleeAttack(cid.incrementAndGet(), userId)
            "stop_move" -> {
                val sightId = js.getString("sight")
                StopMove(
                    cid.incrementAndGet(),
                    userId,
                    js.getInteger("x"),
                    js.getInteger("y"),
                    Direction.valueOf(sightId)
                )
            }
            else -> null
        }

    }


    fun onOp(pkg: UserPackage) {
        val messages = JsonArray()
        pkg.ops().forEach {
            messages.add(JsonMapper.toJson(it as OutOp))
        }

        val clientRes = JsonObject()
            .put("tick", pkg.tick()) //todo hardcode
            .put("time", System.currentTimeMillis() / 1000)
            .put("messages", messages)
        ws.writeTextMessage(clientRes.toString())
    }

    private fun setupClient() {
        send(Login(cid.incrementAndGet(), userId))
    }
}
